### PR TITLE
[Wait for #1798][lstm] Refactoring lstm

### PR DESCRIPTION
**Pull request by lhs8928**
Commit 1:     [lstmcell core] prepare refactoring lstmcell core layer
    
     - LSTM cell core layer will be refactoring from layer based to function based.
       This commit prepare the core functions which is not used right now.
    
    Self evaluation:
    
    Build test: [X]Passed [ ]Failed [ ]Skipped
    Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 2:     [lstmcell] refactoring lstmcell layer
    
     - Refactoring LSTMCell layer to use LSTM core functions.
    
    Self evaluation:
    
    Build test: [X]Passed [ ]Failed [ ]Skipped
    Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 3:     [zoneout lstmcell] refactoring zoneout lstmcell layer
    
     - Refactoring zoneout lstmcell layer to use lstmcore functions.
     - Preserve lstm_cell_state tensor for calcGradient.
     - Remove lstmcell core layer
    
    Self evaluation:
    
    Build test: [X]Passed [ ]Failed [ ]Skipped
    Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 4:     [lstm] refactoring lstm layer
    
     - Refactoring lstm layer to use lstm cell core functions.
    
    Self evaluation:
    
    Build test: [X]Passed [ ]Failed [ ]Skipped
    Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/layers/zoneout_lstmcell.cpp [hunk 1]
```diff
@@ -262,402 +211,335 @@ void ZoneoutLSTMCellLayer::finalize(InitLayerContext &context) {
...
-    cell_state_zoneout_mask.getBatchSlice(timestep, 1);
-  next_cell_state_zoneout_mask.reshape({batch_size, 1, 1, unit});
+  cs_zoneout_mask.reshape({max_timestep, 1, batch_size, unit});
+  Tensor cell_state_zoneout_mask = cs_zoneout_mask.getBatchSlice(timestep, 1);
+  cell_state_zoneout_mask.reshape({batch_size, 1, 1, unit});
   Tensor prev_cell_state_zoneout_mask;
   if (!test) {
     prev_cell_state_zoneout_mask =
-      next_cell_state_zoneout_mask.zoneout_mask(cell_state_zoneout_rate);
+      cell_state_zoneout_mask.zoneout_mask(cell_state_zoneout_rate);
```
Comment (zhoonit) [inline nntrainer/layers/zoneout_lstmcell.cpp:511]: This should not create another zoneout_mask, is this fixed later?

File: nntrainer/layers/zoneout_lstmcell.cpp [hunk 2]
```diff
@@ -262,402 +211,335 @@ void ZoneoutLSTMCellLayer::finalize(InitLayerContext &context) {
...
-    hidden_state_zoneout_mask.getBatchSlice(timestep, 1);
-  next_hidden_state_zoneout_mask.reshape({batch_size, 1, 1, unit});
+  hs_zoneout_mask.reshape({max_timestep, 1, batch_size, unit});
+  Tensor hidden_state_zoneout_mask = hs_zoneout_mask.getBatchSlice(timestep, 1);
+  hidden_state_zoneout_mask.reshape({batch_size, 1, 1, unit});
   Tensor prev_hidden_state_zoneout_mask;
   if (!test) {
     prev_hidden_state_zoneout_mask =
-      next_hidden_state_zoneout_mask.zoneout_mask(hidden_state_zoneout_rate);
+      hidden_state_zoneout_mask.zoneout_mask(hidden_state_zoneout_rate);
```
Comment (zhoonit) [inline nntrainer/layers/zoneout_lstmcell.cpp:489]: same here

**Review by zhoonit — COMMENTED** [review]
nice work :)

File: nntrainer/layers/zoneout_lstmcell.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/layers/zoneout_lstmcell.cpp:511]: Checked that this is fixed at next PR

**Review by zhoonit — APPROVED** [review]

**Pull request merged**